### ISSUE TITLE
Apparent memory corruption for ak::Content::classname()

### ISSUE THREAD

**Issue by raymondEhlers** [scikit-hep/awkward#778]
While investigating #774, I was concerned that there may be some issues with gcc 7.3, so I tried to compile fastjet separately and then build my bindings. In the process, I came across what looks like some memory corruption when trying to call `ak::Content::classname()` as a cross check. It either causes memory corruption or a segfault. I'm fairly certain this isn't system specific because I managed to reproduce it in singularity (specifically, using the docker image `python:3.9-buster`), as well as directly on the system.

Again, falling back on my reproducer repo, this time on the branch [`segfault_name`](https://github.com/raymondEhlers/pybind11-awkward-array-test/tree/segfault_name). Running pytest to just print the name of the array results in a segfault:

```bash
(test_env) Singularity> pytest -l -vv tests
===================================================================================== test session starts =====================================================================================
platform linux -- Python 3.9.2, pytest-6.2.2, py-1.10.0, pluggy-0.13.1 -- /home/rehlers/code/quarantine/pybind11-awkward-array-test-name/test_env/bin/python3
cachedir: .pytest_cache
rootdir: /home/rehlers/code/quarantine/pybind11-awkward-array-test-name
collected 2 items

tests/test_simple.py::test_numpy_array Fatal Python error: Aborted

Current thread 0x00007fbeed68b740 (most recent call first):
  File "/home/rehlers/code/quarantine/pybind11-awkward-array-test-name/tests/test_simple.py", line 6 in test_numpy_array
  File "/home/rehlers/code/quarantine/pybind11-awkward-array-test-name/test_env/lib/python3.9/site-packages/_pytest/python.py", line 183 in pytest_pyfunc_call
  File "/home/rehlers/code/quarantine/pybind11-awkward-array-test-name/test_env/lib/python3.9/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/home/rehlers/code/quarantine/pybind11-awkward-array-test-name/test_env/lib/python3.9/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/home/rehlers/code/quarantine/pybind11-awkward-array-test-name/test_env/lib/python3.9/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/home/rehlers/code/quarantine/pybind11-awkward-array-test-name/test_env/lib/python3.9/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/home/rehlers/code/quarantine/pybind11-awkward-array-test-name/test_env/lib/python3.9/site-packages/_pytest/python.py", line 1641 in runtest
  File "/home/rehlers/code/quarantine/pybind11-awkward-array-test-name/test_env/lib/python3.9/site-packages/_pytest/runner.py", line 162 in pytest_runtest_call
  File "/home/rehlers/code/quarantine/pybind11-awkward-array-test-name/test_env/lib/python3.9/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/home/rehlers/code/quarantine/pybind11-awkward-array-test-name/test_env/lib/python3.9/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/home/rehlers/code/quarantine/pybind11-awkward-array-test-name/test_env/lib/python3.9/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/home/rehlers/code/quarantine/pybind11-awkward-array-test-name/test_env/lib/python3.9/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/home/rehlers/code/quarantine/pybind11-awkward-array-test-name/test_env/lib/python3.9/site-packages/_pytest/runner.py", line 255 in <lambda>
  File "/home/rehlers/code/quarantine/pybind11-awkward-array-test-name/test_env/lib/python3.9/site-packages/_pytest/runner.py", line 311 in from_call
  File "/home/rehlers/code/quarantine/pybind11-awkward-array-test-name/test_env/lib/python3.9/site-packages/_pytest/runner.py", line 254 in call_runtest_hook
  File "/home/rehlers/code/quarantine/pybind11-awkward-array-test-name/test_env/lib/python3.9/site-packages/_pytest/runner.py", line 215 in call_and_report
  File "/home/rehlers/code/quarantine/pybind11-awkward-array-test-name/test_env/lib/python3.9/site-packages/_pytest/runner.py", line 126 in runtestprotocol
  File "/home/rehlers/code/quarantine/pybind11-awkward-array-test-name/test_env/lib/python3.9/site-packages/_pytest/runner.py", line 109 in pytest_runtest_protocol
  File "/home/rehlers/code/quarantine/pybind11-awkward-array-test-name/test_env/lib/python3.9/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/home/rehlers/code/quarantine/pybind11-awkward-array-test-name/test_env/lib/python3.9/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/home/rehlers/code/quarantine/pybind11-awkward-array-test-name/test_env/lib/python3.9/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/home/rehlers/code/quarantine/pybind11-awkward-array-test-name/test_env/lib/python3.9/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/home/rehlers/code/quarantine/pybind11-awkward-array-test-name/test_env/lib/python3.9/site-packages/_pytest/main.py", line 348 in pytest_runtestloop
  File "/home/rehlers/code/quarantine/pybind11-awkward-array-test-name/test_env/lib/python3.9/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/home/rehlers/code/quarantine/pybind11-awkward-array-test-name/test_env/lib/python3.9/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/home/rehlers/code/quarantine/pybind11-awkward-array-test-name/test_env/lib/python3.9/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/home/rehlers/code/quarantine/pybind11-awkward-array-test-name/test_env/lib/python3.9/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/home/rehlers/code/quarantine/pybind11-awkward-array-test-name/test_env/lib/python3.9/site-packages/_pytest/main.py", line 323 in _main
  File "/home/rehlers/code/quarantine/pybind11-awkward-array-test-name/test_env/lib/python3.9/site-packages/_pytest/main.py", line 269 in wrap_session
  File "/home/rehlers/code/quarantine/pybind11-awkward-array-test-name/test_env/lib/python3.9/site-packages/_pytest/main.py", line 316 in pytest_cmdline_main
  File "/home/rehlers/code/quarantine/pybind11-awkward-array-test-name/test_env/lib/python3.9/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/home/rehlers/code/quarantine/pybind11-awkward-array-test-name/test_env/lib/python3.9/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/home/rehlers/code/quarantine/pybind11-awkward-array-test-name/test_env/lib/python3.9/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/home/rehlers/code/quarantine/pybind11-awkward-array-test-name/test_env/lib/python3.9/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/home/rehlers/code/quarantine/pybind11-awkward-array-test-name/test_env/lib/python3.9/site-packages/_pytest/config/__init__.py", line 162 in main
  File "/home/rehlers/code/quarantine/pybind11-awkward-array-test-name/test_env/lib/python3.9/site-packages/_pytest/config/__init__.py", line 185 in console_main
  File "/home/rehlers/code/quarantine/pybind11-awkward-array-test-name/test_env/bin/pytest", line 8 in <module>
Aborted
```

while running the function directly show the memory corruption:

```bash
(test_env) Singularity> python tests/test_simple.py
In function for NumpyArray
. Name: NumpyArray8UU
    Partition each element in `a` around `sep`.

    Calls `str.partition` element-wise.

    For each element in `a`, split the element as the first
    occurrence of `sep`, and return 3 strings containing the part
    before the separator, the separator itself, and the part after
    the separator. If the separator is not found, return 3 strings
    containing the string itself, followed by two empty strings.

    Parameters
    ----------
    a : array_like, {str, unicode}
        Input array
    sep : {str, unicode}
        Separator to split each string element in `a`.

    Returns
    -------
    out : ndarray, {str, unicode}
        Output array of str or unicode, depending on input type.
        The output array will have an extra dimension with 3
        elements per input element.

    See also
    --------
    str.partition

    A8U%8UValidate a pyc against the sourcfree(): invalid pointer
Aborted
```

This may just be a forgotten null, and I don't know if this is related to my underlying issues in #774, but seemed better to report. As always, your help is appreciated.

```python
>>> awkward.__version__
'1.2.0rc2'
```

**Comment by jpivarski** [comment]
That might be what gets reported as the segfault, but I doubt that it's the classname function itself that's at fault:

https://github.com/scikit-hep/awkward-1.0/blob/5999fbd39793a036802b525f34c1dfb3d43547c2/src/libawkward/array/NumpyArray.cpp#L510-L513

(It's not defined with an ampersand or anything.) But that's what's terrible about segfaults—they don't give useful debugging info. When I get a segfault, I bisect the code with deliberate exceptions. Throwing an exception before the segfault tells you that the code up to that point is okay; not getting to the exception because of the segfault is not informative. I keep moving the exception until I find the cause.

In this case, almost nothing is happening in this module. I could try it out Monday morning (or afternoon, or Tuesday, as things get pushed back), but this is the methodology that I'd use.

**Comment by raymondEhlers** [comment]
I didn't yet look into the awkward code base, but you're absolutely right - it seems unlikely to come from the class name itself. I suppose that's what I was hoping for, in that something so simple shouldn't be causing problems, which may make it easier to isolate.

Thanks for the suggestions on debugging strategy. I didn't yet embark on debugging in awkward itself because the code structure is still a bit opaque to me, and I haven't had time to sit down and try to understand it detail. I will try to find some time, but not sure I'll get there in the next day or so. I've tried so many variations and tests trying to resolve #774 that at this point, I'm mainly hoping that it's not due to a stupid mistake in how I'm building the package that I've somehow managed to integrate into my reproducer :-)

In some other tests, working with gdb had been helpful, but here I clearly need to recompile awkward with debug symbols. Just for completeness (it's really not helpful), here is what I get for the bt when calling python directly (ie. not through pytest):

```
(gdb) bt
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x00007fa9f07e3535 in __GI_abort () at abort.c:79
#2  0x00007fa9f083a508 in __libc_message (action=action@entry=do_abort, fmt=fmt@entry=0x7fa9f094528d "%s\n") at ../sysdeps/posix/libc_fatal.c:181
#3  0x00007fa9f0840c1a in malloc_printerr (str=str@entry=0x7fa9f094343b "free(): invalid pointer") at malloc.c:5341
#4  0x00007fa9f084242c in _int_free (av=<optimized out>, p=<optimized out>, have_lock=<optimized out>) at malloc.c:4165
#5  0x00007fa9d6f59a82 in ?? () from /home/rehlers/code/quarantine/pybind11-awkward-array-test-name/pybind11_awkward/_src.cpython-39-x86_64-linux-gnu.so
#6  0x00007fa9d6f639b5 in ?? () from /home/rehlers/code/quarantine/pybind11-awkward-array-test-name/pybind11_awkward/_src.cpython-39-x86_64-linux-gnu.so
#7  0x00007fa9d6f606fe in ?? () from /home/rehlers/code/quarantine/pybind11-awkward-array-test-name/pybind11_awkward/_src.cpython-39-x86_64-linux-gnu.so
#8  0x00007fa9f0c849c3 in cfunction_call (func=0x7fa9d73f1c20, args=<optimized out>, kwargs=<optimized out>) at Objects/methodobject.c:539
#9  0x00007fa9f0c3c58a in _PyObject_MakeTpCall (tstate=0x55757a578170, callable=0x7fa9d73f1c20, args=<optimized out>, nargs=1, keywords=0x0) at Objects/call.c:191
#10 0x00007fa9f0bfb49b in _PyObject_VectorcallTstate (kwnames=<optimized out>, nargsf=<optimized out>, args=<optimized out>, callable=<optimized out>, tstate=<optimized out>)
    at ./Include/cpython/abstract.h:116
#11 _PyObject_VectorcallTstate (kwnames=<optimized out>, nargsf=<optimized out>, args=<optimized out>, callable=<optimized out>, tstate=<optimized out>) at ./Include/cpython/abstract.h:103
#12 PyObject_Vectorcall (kwnames=<optimized out>, nargsf=<optimized out>, args=<optimized out>, callable=<optimized out>) at ./Include/cpython/abstract.h:127
...
```

**Comment by jpivarski** [comment]
Hi! What is the status of this? Is it still an ongoing investigation?

**Comment by raymondEhlers** [comment]
Hi! Unfortunately, this is still an open issue. The current status is the same - I can reproduce it in a container via singularity, but haven't had time to go further. It took me 2-3 days of on and off work on this just to get it isolated, so while I appreciate your debugging suggestions, it's been hard to find more time to investigate when it's moving that slowly. I can perhaps spend part of a day on the weekend, but if I can't sort it out by then, I'll most likely put my time into trying to find a (potentially very hacky) workaround rather than a fix, which I hope will come to a faster resolution

**Comment by raymondEhlers** [comment]
I gave it a shot this weekend, but didn't make any meaningful progress. Reproducing it is still hit or miss, even with the container. So maybe it's compiler specific, or there's some contamination that I don't understand, or I'm introducing a problem somehow in compiling or linking my code, or something else..?

In any case, my plan from here is to try to work around it by dropping passing awkward arrays directly into c++. Instead, I'll just iterate over the arrays in python and pass numpy arrays into c++. It would have been nice to use the existing code since it works on some platforms, but I guess this probably won't be a bottleneck for jet finding, so perhaps it will be just fine.

Thanks for your help. I don't foresee any progress here, so you can close it if you'd like

**Comment by jpivarski** [comment]
Okay, thanks for looking into this. I find that the recommendation of passing pure NumPy arrays through pybind11 is general enough for many situations. For jet-finding, that should be on the order of 4 arrays (5 for jaggedness). The only issue would be if you didn't know the structure, but even then, there would be [ak.to_buffers](https://awkward-array.readthedocs.io/en/latest/_auto/ak.to_buffers.html)/[ak.from_buffers](https://awkward-array.readthedocs.io/en/latest/_auto/ak.from_buffers.html) to turn general arrays into nothing but NumPy, assuming you knew the `from_buffers` protocol well enough to use them on the C++ side.

**Comment by raymondEhlers** [comment]
I'm not sure I fully understand your comment, so let me just ask one question to ensure I don't go too far off base. What I had in mind was something as simple as (schematically):

```python
events = particles = ak.Array({
    "px": [[99.0, 4.0, -99.0], [0.1, -0.1, 0]],
    "py": [[0.1, -0.1, 0], [99.0, 4.0, -99.0]],
    "pz": [[0, 0, 0], [0, 0, 0]],
    "E": [[100.0, 5.0, 99.0], [100.0, 5.0, 99.0]],
})

jets_px = []
for event in events:
    result = my_cpp_func_to_perform_jet_finding(
        np.asarray(event.px),
        np.asarray(event.py),
        np.asarray(event.pz),
        np.asarray(event.E),
    ))
    # Some map of numpy arrays
    jets_px.append(result["px"])
    # And so on for the rest.
# Build the output, presumably based around offsets and np.concatenate each momentum component...
```

Admittedly, I haven't actually written this yet, and I'm not a pybind11 expert, but I was hoping to let it take care of the details of passing arrays in and out. So I didn't foresee any use of to_buffers/from_buffers. Is this too simplistic? (I know this is getting pretty far out of scope for you and your work on awkward, so I won't keep up on this topic after this question. But any advice is greatly appreciated!)

**Comment by jpivarski** [comment]
I was talking in generality: you have a simple enough, fixed structure that you don't need to_buffers/from_buffers. Forget that—I was just saying that even fully general data structures don't _really_ need the C++ interface, something I've been thinking a lot about recently.

One thing about your case: you'll have to flatten the px/py/pz/E before turning it into a NumPy array. You can do it only once with the high-level [ak.flatten](https://awkward-array.readthedocs.io/en/latest/_auto/ak.flatten.html) function, or you can delve into the low-level [ListArray](https://awkward-array.readthedocs.io/en/latest/ak.layout.ListArray.html) or [ListOffsetArray](https://awkward-array.readthedocs.io/en/latest/ak.layout.ListOffsetArray.html)—if you do any slices of the array, you don't know which type it will be, so [ak.flatten](https://awkward-array.readthedocs.io/en/latest/_auto/ak.flatten.html) is easier).

To turn the results into an Awkward Array, you can [ak.unflatten](https://awkward-array.readthedocs.io/en/latest/_auto/ak.unflatten.html) them, given an array of counts, or construct a [ListOffsetArray](https://awkward-array.readthedocs.io/en/latest/ak.layout.ListOffsetArray.html), given an array of offsets.

**Comment by raymondEhlers** [comment]
Thanks for clarifying, and thanks as always for all of the advice!
